### PR TITLE
Implement ZIO.not

### LINKED ISSUE #4197: Implement ZIO.not
In my code base, I found useful having such combinator:

```scala
def not[R, E](effect: ZIO[R, E, Boolean]) =
  effect.map(result => !result)
```

e.g., if you have method `activated(foo)`, it's quite concise to write `not(activated(foo))`.

### PR DESCRIPTION
**Pull request by evis**
Resolves #4197

### CODE REVIEW

**Comment by evis** [comment]
@adamgfraser could you restart builds please? They failed because of reasons not related to this PR.

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 1]
```diff
@@ -3291,6 +3297,12 @@ object ZIO extends ZIOCompanionPlatformSpecific {
     if (cont(initial)) body(initial) *> loop_(inc(initial))(cont, inc)(body)
     else ZIO.unit
 
+  /**
+   * Returns a new effect where boolean value of this effect is negated.
+   */
+  def not[R, E](effect: ZIO[R, E, Boolean]): ZIO[R, E, Boolean] =
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZIO.scala:3303]: Would be good to have variants of this for the type alias companion objects.

**Comment by evis** [comment]
Again, tests failed due to some random timeouts not related to this PR :(

File: core/shared/src/main/scala/zio/IO.scala [hunk 2]
```diff
@@ -675,6 +675,12 @@ object IO {
    */
   val none: UIO[Option[Nothing]] = ZIO.none
 
+  /**
+   *  @see See [[zio.ZIO.not]]
+   */
+  def not[R, E](effect: ZIO[R, E, Boolean]): ZIO[R, E, Boolean] =
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/IO.scala:681]: No `R` type parameter here and just accept and return an `IO[E, A]`. Similar thing for each of the other companion object methods so we return the more specific type.

**Review by adamgfraser — APPROVED** [review]
Thank you!

**Pull request merged**